Public/Sdk/Experimental/NodeJs/npm.dsc

```diff
@@ -41,14 +41,66 @@ namespace Npm {
         };
     }
 
+    @@public
+    export function installFromPackageJson(workingDirectory: StaticDirectory): OpaqueDirectory {
+        const wd = workingDirectory.root;
+        const nodeModulesPath = d`${wd}/node_modules`;
+        const npmCachePath = Context.getNewOutputDirectory('npm-install-cache');
+
+        const arguments: Argument[] = [
+            Cmd.argument(Artifact.input(Node.npmCli)),
+            Cmd.argument("install"),
+            Cmd.option("--cache ", Artifact.none(npmCachePath)), // Forces the npm cache to use this output folder for this object so that it doesn't write to user folder
+        ];
+
+        const result = Node.run({
+            arguments: arguments,
+            workingDirectory: wd,
+            dependencies: [ workingDirectory ].filter(x => x !== undefined),
+            outputs: [
+                { directory: wd, kind: "shared" },

```

Comment:
Clarification: here we declare the entire `wd` to be a "shared opaque directory" because `npm install` will write files directly in it (e.g., the `package-lock.json` file).

If `package-lock.json` were the only file written by `npm install` to the root of `wd`, that would be great: in that case, we could declare that exact file as its output and be done with it.  The problem is, however, that `npm install` first writes a file like `package-lock.json.<some-random-number>` and then renames that file to `package-lock.json`.  Since we don't know what `<some-random-number>` is going to be from run to run, we basically have to declare the entire folder as a shared opaque output folder.